File: app/src/main/java/com/krystianwsul/checkme/domainmodel/local/LocalInstance.java
package com.krystianwsul.checkme.domainmodel.local;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.krystianwsul.checkme.domainmodel.DomainFactory;
import com.krystianwsul.checkme.domainmodel.Instance;
import com.krystianwsul.checkme.persistencemodel.InstanceRecord;
import com.krystianwsul.checkme.utils.CustomTimeKey;
import com.krystianwsul.checkme.utils.TaskKey;
import com.krystianwsul.checkme.utils.time.Date;
import com.krystianwsul.checkme.utils.time.DateTime;
import com.krystianwsul.checkme.utils.time.ExactTimeStamp;
import com.krystianwsul.checkme.utils.time.NormalTime;
import com.krystianwsul.checkme.utils.time.Time;
import com.krystianwsul.checkme.utils.time.TimePair;

import junit.framework.Assert;

public class LocalInstance extends Instance {
    @Nullable
    private InstanceRecord mInstanceRecord;

    @Nullable
    private Integer mTaskId;

    @Nullable
    private DateTime mScheduleDateTime;

    LocalInstance(@NonNull DomainFactory domainFactory, @NonNull InstanceRecord instanceRecord) {
        super(domainFactory);

        mInstanceRecord = instanceRecord;

        mTaskId = null;
        mScheduleDateTime = null;
    }

    public LocalInstance(@NonNull DomainFactory domainFactory, int taskId, @NonNull DateTime scheduleDateTime) {
        super(domainFactory);

        mInstanceRecord = null;

        mTaskId = taskId;
        mScheduleDateTime = scheduleDateTime;
    }

    public int getTaskId() {
        if (mInstanceRecord != null) {
            Assert.assertTrue(mTaskId == null);
            Assert.assertTrue(mScheduleDateTime == null);

            return mInstanceRecord.getTaskId();
        } else {
            Assert.assertTrue(mTaskId != null);
            Assert.assertTrue(mScheduleDateTime != null);

            return mTaskId;
        }
    }

    @NonNull
    @Override
    public TaskKey getTaskKey() {
        return new TaskKey(getTaskId());
    }

    @NonNull
    @Override
    public String getName() {
        return getTask().getName();
    }

    @NonNull
    @Override
    public Date getScheduleDate() {
        if (mInstanceRecord != null) {
            Assert.assertTrue(mTaskId == null);
            Assert.assertTrue(mScheduleDateTime == null);

            return new Date(mInstanceRecord.getScheduleYear(), mInstanceRecord.getScheduleMonth(), mInstanceRecord.getScheduleDay());
        } else {
            Assert.assertTrue(mTaskId != null);
            Assert.assertTrue(mScheduleDateTime != null);

            return mScheduleDateTime.getDate();
        }
    }

    @NonNull
    @Override
    protected Time getScheduleTime() {
        if (mInstanceRecord != null) {
            Assert.assertTrue(mTaskId == null);
            Assert.assertTrue(mScheduleDateTime == null);

            Integer customTimeId = mInstanceRecord.getScheduleCustomTimeId();
            Integer hour = mInstanceRecord.getScheduleHour();
            Integer minute = mInstanceRecord.getScheduleMinute();

            Assert.assertTrue((hour == null) == (minute == null));
            Assert.assertTrue((customTimeId == null) != (hour == null));

            if (customTimeId != null) {
                return mDomainFactory.getCustomTime(new CustomTimeKey(customTimeId));
            } else {
                return new NormalTime(hour, minute);
            }
        } else {
            Assert.assertTrue(mTaskId != null);
            Assert.assertTrue(mScheduleDateTime != null);

            return mScheduleDateTime.getTime();
        }
    }

    @NonNull
    @Override
    public Date getInstanceDate() {
        if (mInstanceRecord != null) {
            Assert.assertTrue(mTaskId == null);
            Assert.assertTrue(mScheduleDateTime == null);

            Assert.assertTrue((mInstanceRecord.getInstanceYear() == null) == (mInstanceRecord.getInstanceMonth() == null));
            Assert.assertTrue((mInstanceRecord.getInstanceYear() == null) == (mInstanceRecord.getInstanceDay() == null));
            if (mInstanceRecord.getInstanceYear() != null)
                return new Date(mInstanceRecord.getInstanceYear(), mInstanceRecord.getInstanceMonth(), mInstanceRecord.getInstanceDay());
            else
                return getScheduleDate();
        } else {
            Assert.assertTrue(mTaskId != null);
            Assert.assertTrue(mScheduleDateTime != null);

            return mScheduleDateTime.getDate();
        }
    }

    @NonNull
    @Override
    protected Time getInstanceTime() {
        if (mInstanceRecord != null) {
            Assert.assertTrue(mTaskId == null);
            Assert.assertTrue(mScheduleDateTime == null);

            Assert.assertTrue((mInstanceRecord.getInstanceHour() == null) == (mInstanceRecord.getInstanceMinute() == null));
            Assert.assertTrue((mInstanceRecord.getInstanceHour() == null) || (mInstanceRecord.getInstanceCustomTimeId() == null));

            if (mInstanceRecord.getInstanceCustomTimeId() != null) {
                return mDomainFactory.getCustomTime(new CustomTimeKey(mInstanceRecord.getInstanceCustomTimeId()));
            } else if (mInstanceRecord.getInstanceHour() != null) {
                return new NormalTime(mInstanceRecord.getInstanceHour(), mInstanceRecord.getInstanceMinute());
            } else {
                return getScheduleTime();
            }
        } else {
            Assert.assertTrue(mTaskId != null);
            Assert.assertTrue(mScheduleDateTime != null);

            return mScheduleDateTime.getTime();
        }
    }

    @Override
    public void setInstanceDateTime(@NonNull Date date, @NonNull TimePair timePair, @NonNull ExactTimeStamp now) {
        Assert.assertTrue(isRootInstance(now));

        if (mInstanceRecord == null)
            createInstanceHierarchy(now);

        mInstanceRecord.setInstanceYear(date.getYear());
        mInstanceRecord.setInstanceMonth(date.getMonth());
        mInstanceRecord.setInstanceDay(date.getDay());

        if (timePair.mCustomTimeKey != null) {
            Assert.assertTrue(timePair.mHourMinute == null);
            Assert.assertTrue(timePair.mCustomTimeKey.mLocalCustomTimeId != null);
            Assert.assertTrue(TextUtils.isEmpty(timePair.mCustomTimeKey.mRemoteCustomTimeId));

            mInstanceRecord.setInstanceCustomTimeId(timePair.mCustomTimeKey.mLocalCustomTimeId);
            mInstanceRecord.setInstanceHour(null);
            mInstanceRecord.setInstanceMinute(null);
        } else {
            Assert.assertTrue(timePair.mHourMinute != null);

            mInstanceRecord.setInstanceCustomTimeId(null);
            mInstanceRecord.setInstanceHour(timePair.mHourMinute.getHour());
            mInstanceRecord.setInstanceMinute(timePair.mHourMinute.getMinute());
        }

        mInstanceRecord.setNotified(false);
    }

    @Nullable
    @Override
    public ExactTimeStamp getDone() {
        if (mInstanceRecord == null)
            return null;

        Long done = mInstanceRecord.getDone();
        if (done != null)
            return new ExactTimeStamp(done);
        else
            return null;
    }

    @Override
    public void setDone(boolean done, @NonNull ExactTimeStamp now) {
        if (done) {
            if (mInstanceRecord == null) {
                createInstanceHierarchy(now);
                mInstanceRecord.setDone(now.getLong());
            } else {
                mInstanceRecord.setDone(now.getLong());
            }
        } else {
            Assert.assertTrue(mInstanceRecord != null);
            mInstanceRecord.setDone(null);
            mInstanceRecord.setNotified(false);
        }
    }

    @Override
    public void createInstanceHierarchy(@NonNull ExactTimeStamp now) {
        Assert.assertTrue((mInstanceRecord == null) != (mScheduleDateTime == null));
        Assert.assertTrue((mTaskId == null) == (mScheduleDateTime == null));

        Instance parentInstance = getParentInstance(now);
        if (parentInstance != null)
            parentInstance.createInstanceHierarchy(now);

        if (mInstanceRecord == null)
            createInstanceRecord(now);
    }

    private void createInstanceRecord(@NonNull ExactTimeStamp now) {
        LocalTask localTask = (LocalTask) getTask();

        DateTime scheduleDateTime = getScheduleDateTime();

        mTaskId = null;
        mScheduleDateTime = null;

        mInstanceRecord = mDomainFactory.getLocalFactory().createInstanceRecord(localTask, this, scheduleDateTime, now);
    }

    @Deprecated
    public long getHierarchyTime() {
        Assert.assertTrue(mInstanceRecord != null);

        return mInstanceRecord.getHierarchyTime();
    }

    @Override
    public boolean getNotified() {
        return (mInstanceRecord != null && mInstanceRecord.getNotified());
    }

    @Override
    public void setNotified(@NonNull ExactTimeStamp now) {
        if (mInstanceRecord == null)
            createInstanceHierarchy(now);

        Assert.assertTrue(mInstanceRecord != null);
        mInstanceRecord.setNotified(true);
    }

    @Override
    public boolean getNotificationShown() {
        return (mInstanceRecord != null && mInstanceRecord.getNotificationShown());
    }

    @Override
    public void setNotificationShown(boolean notificationShown, @NonNull ExactTimeStamp now) {
        if (mInstanceRecord == null)
            createInstanceHierarchy(now);

        Assert.assertTrue(mInstanceRecord != null);

        mInstanceRecord.setNotificationShown(notificationShown);
    }

    @Override
    public boolean exists() {
        Assert.assertTrue((mInstanceRecord == null) != (mScheduleDateTime == null));
        Assert.assertTrue((mTaskId == null) == (mScheduleDateTime == null));

        return (mInstanceRecord != null);
    }

    @Override
    public void delete() {
        Assert.assertTrue(mInstanceRecord != null);

        mDomainFactory.getLocalFactory().deleteInstance(this);

        mInstanceRecord.delete();
    }
}

File: app/src/main/java/com/krystianwsul/checkme/domainmodel/local/LocalInstance.kt
package com.krystianwsul.checkme.domainmodel.local

import com.krystianwsul.checkme.domainmodel.DomainFactory
import com.krystianwsul.checkme.domainmodel.Instance
import com.krystianwsul.checkme.firebase.RemoteProject
import com.krystianwsul.checkme.persistencemodel.InstanceShownRecord
import com.krystianwsul.checkme.persistencemodel.LocalInstanceRecord
import com.krystianwsul.checkme.utils.CustomTimeKey
import com.krystianwsul.checkme.utils.InstanceData
import com.krystianwsul.checkme.utils.InstanceData.VirtualInstanceData
import com.krystianwsul.checkme.utils.time.Date
import com.krystianwsul.checkme.utils.time.DateTime
import com.krystianwsul.checkme.utils.time.ExactTimeStamp
import com.krystianwsul.checkme.utils.time.TimePair


class LocalInstance : Instance {

    override var instanceData: InstanceData<Int, LocalInstanceRecord>

    val taskId
        get() = instanceData.let {
            when (it) {
                is InstanceData.RealInstanceData<Int, LocalInstanceRecord> -> it.instanceRecord.taskId
                is VirtualInstanceData<Int, LocalInstanceRecord> -> it.taskId
            }
        }

    override val notified get() = (instanceData as? InstanceData.RealInstanceData<Int, LocalInstanceRecord>)?.instanceRecord?.notified == true

    override val notificationShown get() = (instanceData as? InstanceData.RealInstanceData<Int, LocalInstanceRecord>)?.instanceRecord?.notificationShown == true

    override val scheduleCustomTimeKey
        get() = instanceData.let {
            when (it) {
                is InstanceData.RealInstanceData<Int, LocalInstanceRecord> -> it.instanceRecord
                        .scheduleCustomTimeId
                        ?.let { CustomTimeKey.LocalCustomTimeKey(it) }
                is VirtualInstanceData<Int, LocalInstanceRecord> -> it.scheduleDateTime
                        .time
                        .timePair
                        .customTimeKey
            }
        }

    override val task get() = domainFactory.localFactory.getTaskForce(taskId)

    override val remoteNullableProject: RemoteProject? = null

    override val remoteNonNullProject get() = throw UnsupportedOperationException()

    override val remoteCustomTimeKey: Pair<String, String>? = null

    override val nullableInstanceShownRecord: InstanceShownRecord? = null

    constructor(domainFactory: DomainFactory, localInstanceRecord: LocalInstanceRecord) : super(domainFactory) {
        instanceData = LocalRealInstanceData(localInstanceRecord)
    }

    constructor(domainFactory: DomainFactory, taskId: Int, scheduleDateTime: DateTime) : super(domainFactory) {
        instanceData = VirtualInstanceData(taskId, scheduleDateTime)
    }

    override fun setInstanceDateTime(date: Date, timePair: TimePair, now: ExactTimeStamp) {
        check(isRootInstance(now))

        createInstanceHierarchy(now)

        (instanceData as LocalRealInstanceData).instanceRecord.let {
            it.instanceYear = date.year
            it.instanceMonth = date.month
            it.instanceDay = date.day

            val (customTimeId, hour, minute) = timePair.destructureLocal()
            it.instanceCustomTimeId = customTimeId
            it.instanceHour = hour
            it.instanceMinute = minute

            it.notified = false
        }
    }

    override fun setDone(done: Boolean, now: ExactTimeStamp) {
        if (done) {
            createInstanceHierarchy(now)

            (instanceData as LocalRealInstanceData).instanceRecord.let {
                it.done = now.long
                it.notified = false
            }
        } else {
            (instanceData as LocalRealInstanceData).instanceRecord.done = null
        }
    }

    override fun createInstanceRecord(now: ExactTimeStamp) {
        val localTask = task

        instanceData = LocalRealInstanceData(domainFactory.localFactory.createInstanceRecord(localTask, this, scheduleDate, scheduleTimePair, now))
    }

    override fun setNotified(now: ExactTimeStamp) {
        createInstanceHierarchy(now)

        (instanceData as LocalRealInstanceData).instanceRecord.notified = true
    }

    override fun setNotificationShown(notificationShown: Boolean, now: ExactTimeStamp) {
        createInstanceHierarchy(now)

        (instanceData as LocalRealInstanceData).instanceRecord.notificationShown = notificationShown
    }

    override fun delete() {
        check(instanceData is LocalRealInstanceData)

        domainFactory.localFactory.deleteInstance(this)

        (instanceData as LocalRealInstanceData).instanceRecord.delete()
    }

    override fun belongsToRemoteProject() = false

    override fun getNullableOrdinal() = (instanceData as? LocalRealInstanceData)?.instanceRecord?.ordinal

    private inner class LocalRealInstanceData(localInstanceRecord: LocalInstanceRecord) : InstanceData.RealInstanceData<Int, LocalInstanceRecord>(localInstanceRecord) {

        override fun getCustomTime(customTimeId: Int) = domainFactory.getCustomTime(CustomTimeKey.LocalCustomTimeKey(customTimeId))

        override fun getSignature() = name + " " + instanceKey.toString()
    }
}
